Add definition for 'invalid' objects, and some editing

kainino0x opened the pull request:
<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/kainino0x/gpuweb/pull/528.html" title="Last updated on Jan 8, 2020, 11:39 PM UTC (205febf)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/gpuweb/gpuweb/528/70be08a...kainino0x:205febf.html" title="Last updated on Jan 8, 2020, 11:39 PM UTC (205febf)">Diff</a>

Kangz commented on spec/index.bs:
```diff
@@ -141,10 +144,21 @@ WebGPU defines the lowest allowed limits and capabilities of any {{GPUAdapter}}.
+## Dot Syntax ## {#dot-syntax}
+
+In this specification, the `.` ("dot") syntax, common in programming languages, is used.
+The phrasing "`Foo.Bar`" means "the `Two` member of the value (or interface) `Foo`."
```
nit: `Two` should be `Bar`

Kangz commented on spec/index.bs:
```diff
@@ -141,10 +144,21 @@ WebGPU defines the lowest allowed limits and capabilities of any {{GPUAdapter}}.
+Terminology & Conventions {#terminology-and-conventions}
+========================================================
+
+## Dot Syntax ## {#dot-syntax}
```
Thank you for making this, it will be extremely handy :)

Kangz commented on spec/index.bs:
```diff
@@ -154,70 +168,31 @@ WebGPU's coordinate systems match DirectX and Metal's coordinate systems in grap
+<div class=note>
+    [=Invalid=] objects result from a number of situations, including:
+
+      - If there is an error in the creation of the object, it is immediately invalid.
```
Does encoding a bad command on an encoder make it invalid?

JusSn replied:
This would be better reserved for that section of the spec, but good question.

kainino0x replied:
Agree, not necessary to talk about it here since this is non-comprehensive and non-normative.

But probably yes? It doesn't make much difference between that and "internal field that says whether it's going to be invalid"

Kangz commented on spec/index.bs:
```diff
@@ -231,25 +206,46 @@ interface mixin GPUObjectBase {
+        A label which can be used by development tools (such as error/warning messages,
+        browser developer tools, or platform debugging utilities) to identify the underlying
+        [=internal object=] to the developer.
+        It is not machine-readable.
```
What does this mean?

kainino0x replied:
"It has no specified format, and therefore cannot be reliably machine-parsed."?

Kangz commented on spec/index.bs:
```diff
@@ -317,7 +314,7 @@ different browser codepaths or system drivers on the same physical adapters.
 (even if some are [=invalid=]),
 and two of these could refer to different instances of the same physical
-configuration (e.g. if the GPU reset, or were disconnected and reconnected).
+configuration (e.g. if the GPU was reset or disconnected & reconnected).
```
Why the `&` shorthand?

Kangz commented on spec/index.bs:
```diff
@@ -429,9 +428,7 @@ through which [=internal objects=] are created.
-when the [=device=] is lost, all objects created from it become invalid.
-
-<!-- TODO: write and link to spec about what invalid means -->
+when the [=device=] is lost, it and all objects created from it become [=invalid=].
```
nit: also recursively for example for texture views.

JusSn commented on spec/index.bs:
```diff
@@ -361,34 +358,36 @@ interface GPUAdapter {
 To get a {{GPUAdapter}}, use {{GPU/requestAdapter()}}.
 
+<dfn dictionary>GPURequestAdapterOptions</dfn>
+provides hints to the user agent indicating what
```
```suggestion
Provides hints to the user agent indicating what
```

kainino0x replied:
"GPURequestAdapterOptions" is the beginning of the sentence here.

kainino0x commented on spec/index.bs:
```diff
@@ -141,10 +144,21 @@ WebGPU defines the lowest allowed limits and capabilities of any {{GPUAdapter}}.
+The phrasing "`Foo.Bar`" means "the `Two` member of the value (or interface) `Foo`."
+
+For example, where `buffer` is a {{GPUBuffer}}, `buffer.[[device]].[[adapter]]` means
+"the `[[adapter]]` internal slot of the [=device=] contained in the `[[device]]`
```
There is a `[=device=]` contained in the internal slot `[[device]]`, I think that's approximately what it says here. But I'll reorder the words a bit.

kainino0x commented on spec/index.bs:
```diff
@@ -305,9 +301,10 @@ An <dfn dfn>adapter</dfn> represents an implementation of WebGPU on the system.
-If an [=adapter=] becomes unavailable, it becomes <dfn dfn>invalid</dfn>.
+If an [=adapter=] becomes unavailable, it becomes [=invalid=].
 Once invalid, it never becomes valid again.
+Any [=devices=] on the adapter, and [=internal objects=] owned by those devices,
```
> A device is the exclusive owner of all internal objects created from it: when the device is lost, all objects created from it become invalid.

But no, no definition of the concept of "owning".

kainino0x commented on spec/index.bs:
```diff
@@ -305,9 +301,10 @@ An <dfn dfn>adapter</dfn> represents an implementation of WebGPU on the system.
 
-<!--TODO: Move definition of "invalid" somewhere more generic.-->
-If an [=adapter=] becomes unavailable, it becomes <dfn dfn>invalid</dfn>.
+If an [=adapter=] becomes unavailable, it becomes [=invalid=].
```
Added:

```
As a matter of convention, if a [=WebGPU interface=] is referred to as [=invalid=],
it means that the [=internal object=] it encapsulates is [=invalid=].
```

kainino0x commented:
Addressed the comments, here's the diff:
https://github.com/gpuweb/gpuweb/pull/528/files/7658f87d950c35185019eff1b1f6b1f44c95f9d6..7cc72832257bfb74b4a64e2ef7de6c915aa2bb98

Plenty of LGTMs, so landing now. thanks!
